compiler/rustc_hir_typeck/src/pat.rs

```diff
@@ -1881,17 +1891,18 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         let len = unmentioned_fields.len();
         let (prefix, postfix, sp) = match fields {
             [] => match &pat.kind {
-                PatKind::Struct(path, [], false) => {
+                PatKind::Struct(path, [], false) if path.span().eq_ctxt(pat.span) => {
                     (" { ", " }", path.span().shrink_to_hi().until(pat.span.shrink_to_hi()))
```

Comment:
You can generalize this by looking at `path.span().ancestor_within_span(pat.span)`.